webapp/STYLE_GUIDE.md

```diff
@@ -0,0 +1,157 @@
+
+## Side Effects & Lifecycle
+
+- **Rules of Hooks**: Always follow the Rules of Hooks (as enforced by ESLint).
+- **Data Fetching**: Use thunks for data shared across views; reserve fetch inside component effects for local-state-only data.
+- **Thunk Responses**: Always destructure thunk responses into {data, error} variables.
+
+## Event Handling & Accessibility
+
+- **Keyboard Shortcuts**: Use `isKeyPressed(event, Constants.KeyCodes.KEY_NAME)` instead of raw key codes to support different keyboard layouts.
+- **Platform Keys**: Use `cmdOrCtrlPressed(event)` for cross-platform Cmd/Ctrl detection.
+- **Input Protection**: Check if `event.target` is an input/textarea/select before handling shortcuts.
+- **Semantic HTML**: Always use `<button type="button">` for clickable actions, never divs with onClick.
+- **ARIA Labels**: Provide `aria-label` on icon-only buttons and add `aria-hidden="true"` to decorative icons.
+- **Focus Management**: Use `focusElement(elementOrId)` from `utils/a11y_utils` for programmatic focus. Store origin with `a11yController.storeOriginElement()` before opening modals and restore with `a11yController.restoreOriginFocus()` on close.
+- **Modals**: Use `role="dialog"`, `aria-modal="true"`, implement focus trap, and add `a11y__modal` class to disable global navigation.
+
+## A11y Controller (Keyboard Navigation)
+
+- **Purpose**: Enables hierarchical keyboard navigation (F6/Ctrl+~ between regions, arrows within sections).
+- **Regions**: Add `a11y__region` class and `data-a11y-sort-order` attribute to major UI areas for F6 navigation.
+- **Sections**: Add `a11y__section` class to navigable list items for arrow key navigation when region is active.
+- **Modal/Popup Disable**: Add `a11y__modal` or `a11y__popup` class to temporarily disable global navigation.
+- **Additional Attributes**: Use `data-a11y-focus-child`, `data-a11y-loop-navigation`, or `data-a11y-order-reversed` as needed.
+
+## Security Patterns
+
+- **Auth**: Do not manually add Authorization headers; call Client4.setToken during login.
+- **File Uploads**: Pass FormData objects directly to Client4 so the browser sets the correct Content-Type boundary.
+
+## Component Libraries & Packages
+
+- **Search First**: Before creating a new UI component, search for an existing one (e.g., GenericModal, Input) that can accomplish the task.
+- **React Bootstrap**: Avoid using React Bootstrap. Use GenericModal from @mattermost/components instead of React Bootstrap's Modal.
+- **MUI**: If using MUI is necessary, wrap the components in an interface to hide implementation details (as done in components/menu).
+- **Popovers**: Use WithTooltip for simple tooltips. Prefer Floating UI for other popover/menu uses.
+- **Deprecated Packages**: @mattermost/compass-components is deprecated.
+- **Icons**: Prefer using icons from @mattermost/compass-icons/components.
+
+## Styling & Theming (SCSS)
```

Comment:
Should probably call out: Don't add or increase usage of Styled Components unless it's already predominately used in a given context.